step-capstone/src/scripts/Suggestion.js

```diff
@@ -0,0 +1,148 @@
+    return dateObject.getFullYear() + "-" + dateObject.getMonth() + "-" + dateObject.getDate();
+}
+
+const contains = (startTime, endTime, timePoint) => {
+    // return whether timeRange from startTime to endTime contains timePoint
+    return startTime < timePoint && timePoint < endTime;
+}
+
+const overlaps = (openingTime, closingTime, startTime, endTime) => {
+    /**
+     * Case 1: 
+     * opening-closing: |----------|
+     * start-end      :     |----------|
+     * 
+     * Case 2:
+     * opening-closing:         |----------|
+     * start-end      :     |----------| 
+     * 
+     * Case 3:
+     * opening-closing:         |----------|
+     * start-end      :            |------| 
+     * 
+     * Case 4:
+     * opening-closing:         |----|
+     * start-end      :     |----------| 
+     * 
+     */
+    if (contains(openingTime, closingTime, startTime) && !contains(openingTime, closingTime, endTime)) {
+        return millisToMinutes(closingTime - startTime);
+    }
+    else if (contains(startTime, endTime, openingTime) && !contains(startTime, endTime, closingTime)) {
+        return millisToMinutes(endTime - openingTime);
+    }
+    else if (contains(openingTime, closingTime, startTime) && contains(openingTime, closingTime, endTime)) {
+        return millisToMinutes(endTime - startTime);
+    }
+    else if (contains(startTime, endTime, openingTime) && !contains(startTime, endTime, closingTime)) {
+        return millisToMinutes(closingTime - openingTime);
+    } 
+    else {
+        return 0;
+    }
+}
+
+const millisToMinutes = (millis) => {
+  var minutes = Math.floor(millis / 60000);
+  return minutes;
+}
+
+const filterByTimeRange = (results, timeRange) => {
+    results.filter((result) => {
+        if (result.hasOwnProperty("opening_hours")) {
+            let day = timeRange[0].getDay();
+            
+            let openHoursMinutes = result.opening_hour.period[day].open.time;
+            let closeHoursMinutes = (result.opening_hour.period[day].close !== undefined)? 
+                                    result.opening_hour.period[day].close.time
+                                    : "2359";
+            
+
+            let date = getDateString(timeRange[0]);
+            let openingTime = new Date(date + "T" + openHoursMinutes.slice(0, 2) + ":" + openHoursMinutes.splice(2) +":00");
+            let closingTime = new Date(date + "T" + closeHoursMinutes.slice(0, 2) + ":" + closeHoursMinutes.splice(2) + ":00");
+            
+            return overlaps(openingTime, closingTime, timeRange[0], timeRange[1]) >= 45;
+        } 
+        return true;
+    })
+}
+
+const query = (coordinates, service) => {
+    // return results: a map with key : place_id, value: PlaceObject
+    let places = queryPlaces(coordinates, radius, service, ["tourist_attraction", "natural_feature"])
+    places.then(results => {
+        return results;
+    })
+}
+
+const countCat = (placeObject, userCat) => {
```

Comment:
Nit: I find the use of `cat` in your naming a little confusing, since it's often used as an abbreviation for concatenate, not category. However, it's used in a lot of places, so might not be worth changing.

Reply:
Marked as resolved in https://github.com/googleinterns/step98-2020/pull/41.